### PR TITLE
8264897: [lworld] C2: Use BarrierSetC2::obj_allocate to buffer inline type in PhaseMacroExpand::expand_mh_intrinsic_return

### PR DESCRIPTION
**Pull request by kelthuzadx**
Try to reuse existing BarrierSetC2::obj_allocate functionality for PhaseMacroExpand::expand_mh_intrinsic_return.

Testing: x86_64 + linux+ slowdebug/release
[x] test/hotspot/jtreg/runtime/valhalla 
[x] test/hotspot/jtreg/compiler/valhalla 

Thanks,
Yang

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8264897](https://bugs.openjdk.java.net/browse/JDK-8264897): [lworld] C2: Use BarrierSetC2::obj_allocate to buffer inline type in PhaseMacroExpand::expand_mh_intrinsic_return

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - Committer) ⚠️ Review applies to 5e88e8a08c4532840c66177e0032953ba6e2956c

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/385/head:pull/385` \
`$ git checkout pull/385`

Update a local copy of the PR: \
`$ git checkout pull/385` \
`$ git pull https://git.openjdk.java.net/valhalla pull/385/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 385`

View PR using the GUI difftool: \
`$ git pr show -t 385`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/385.diff">https://git.openjdk.java.net/valhalla/pull/385.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/macro.cpp [hunk 1]
```diff
@@ -3232,11 +3217,9 @@ bool PhaseMacroExpand::expand_macro_nodes() {
     case Node::Class_CallStaticJava:
       expand_mh_intrinsic_return(n->as_CallStaticJava());
       C->remove_macro_node(n);
-      assert(C->macro_count() == (old_macro_count - 1), "expansion must have deleted one node from macro list");
       break;
     case Node::Class_FlatArrayCheck:
       expand_flatarraycheck_node(n->as_FlatArrayCheck());
-      assert(C->macro_count() == (old_macro_count - 1), "expansion must have deleted one node from macro list");
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/macro.cpp:3239]: Why did you delete these asserts?

File: src/hotspot/share/opto/macro.cpp [hunk 2]
```diff
@@ -2738,49 +2739,56 @@ void PhaseMacroExpand::expand_mh_intrinsic_return(CallStaticJavaNode* call) {
...
-  Node* old_top = NULL;
-  Node* new_top = NULL;
-  if (UseTLAB) {
-    Node* end_adr = NULL;
-    set_eden_pointers(top_adr, end_adr);
-    Node* end = make_load(ctl, mem, end_adr, 0, TypeRawPtr::BOTTOM, T_ADDRESS);
-    old_top = new LoadPNode(ctl, mem, top_adr, TypeRawPtr::BOTTOM, TypeRawPtr::BOTTOM, MemNode::unordered);
-    transform_later(old_top);
+  // Try to allocate a new buffered inline instance either from TLAB or eden space
+  Node* needgc_ctrl; // needgc means slowcase, i.e. allocation failed
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/macro.cpp:2745]: Should be initialized to `NULL`.

File: src/hotspot/share/opto/macro.cpp [hunk 3]
```diff
@@ -2804,67 +2812,43 @@ void PhaseMacroExpand::expand_mh_intrinsic_return(CallStaticJavaNode* call) {
...
+    fast_ctl = transform_later(new ProjNode(handler_call, TypeFunc::Control));
+    Node* rawmem = transform_later(new ProjNode(handler_call, TypeFunc::Memory));
+    fast_res = transform_later(new ProjNode(handler_call, TypeFunc::Parms));
+    fast_mem = MergeMemNode::make(mem);
+    fast_mem->set_memory_at(Compile::AliasIdxRaw, rawmem);
+    transform_later(fast_mem);
+  }
 
-  Node* r = new RegionNode(4);
+  Node* r = new RegionNode(alloc_in_place ? 4: 3);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/macro.cpp:2848]: Whitespace missing, should be `? 4 : 3`

File: src/hotspot/share/opto/macro.cpp [hunk 4]
```diff
@@ -2804,67 +2812,43 @@ void PhaseMacroExpand::expand_mh_intrinsic_return(CallStaticJavaNode* call) {
...
-  Node* handler_ctl = transform_later(new ProjNode(handler_call, TypeFunc::Control));
-  rawmem = transform_later(new ProjNode(handler_call, TypeFunc::Memory));
-  Node* slowpath_false_res = transform_later(new ProjNode(handler_call, TypeFunc::Parms));
-
-  MergeMemNode* slowpath_false_mem = MergeMemNode::make(mem);
-  slowpath_false_mem->set_memory_at(Compile::AliasIdxRaw, rawmem);
-  transform_later(slowpath_false_mem);
+  Node* fast_ctl;
+  Node* fast_res;
+  MergeMemNode* fast_mem;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/macro.cpp:2838]: Should be initialized to `NULL`.

File: src/hotspot/share/opto/macro.cpp [hunk 5]
```diff
@@ -2804,67 +2812,43 @@ void PhaseMacroExpand::expand_mh_intrinsic_return(CallStaticJavaNode* call) {
...
-  handler_call->init_req(TypeFunc::Parms, pack_handler);
-  handler_call->init_req(TypeFunc::Parms+1, old_top);
-
   // We don't know how many values are returned. This assumes the
   // worst case, that all available registers are used.
   for (uint i = TypeFunc::Parms+1; i < domain->cnt(); i++) {
     if (domain->field_at(i) == Type::HALF) {
       slow_call->init_req(i, top());
-      handler_call->init_req(i+1, top());
+      if (alloc_in_place) handler_call->init_req(i+1, top());
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/macro.cpp:2824]: Please use parentheses for the if body. Same below.

**Review by TobiHartmann — APPROVED** [review]
Nice cleanup! I've added some minor comments, otherwise looks good to me.

**Comment by kelthuzadx** [comment]
Thank you Tobias! Sorry for so many nits, I will check them more carefully in later PRs.

Best Regards,
Yang

File: src/hotspot/share/opto/macro.cpp (hunk 1 above)
Comment (kelthuzadx) [reply src/hotspot/share/opto/macro.cpp:3239]: It seems the checking happens nevertheless right after switch:

https://github.com/openjdk/valhalla/blob/5e88e8a08c4532840c66177e0032953ba6e2956c/src/hotspot/share/opto/macro.cpp#L3220-L3227

Comment (TobiHartmann) [reply src/hotspot/share/opto/macro.cpp:3239]: Right, good catch!
